### PR TITLE
[DGL-Go] Inference for Graph Prediction Pipeline

### PR DESCRIPTION
**Pull request by mufeili**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    model.eval()
+    all_pred = []
+
+    with torch.no_grad():
+        for _, (g, labels) in enumerate(tqdm(loader, desc="Iteration")):
+            g = g.to(device)
+            node_feat = g.ndata['feat']
+            edge_feat = g.edata['feat']
+            pred = model(g, node_feat, edge_feat)
+            pred = (pred.sigmoid() >= 0.5).long()
```
Comment (jermainewang) [inline dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:23]: Look like we only support binary classification here. Probably should add a check in the training script.

File: dglgo/dglgo/apply_pipeline/graphpred/gen.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+pipeline_comments = {
+    "batch_size": "Graph batch size",
+    "num_workers": "Number of workers for data loading",
+    "save_path": "Directory to save the inference results"
+}
+
+class ApplyGraphpredPipelineCfg(BaseModel):
+    batch_size: int = 32
+    num_workers: int = 4
+    save_path: str = "apply_results"
```
[the remark this answers is not in the document]
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/gen.py:23]: This is mostly for extension purpose, which might include more results like evaluation metrics.

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    # validation
+    train_data_name = cfg['model']['data_info']['name']
+    infer_data_name = cfg['data']['name']
+    if train_data_name.startswith('ogbg-mol'):
+        assert infer_data_name.startswith('ogbg-mol'), 'Expect the inference data name to start \
+            with ogbg-mol, got {}'.format(infer_data_name)
+    else:
+        assert train_data_name == infer_data_name, 'Expect the training and inference data to \
+            have the same name, got {} and {}'.format(train_data_name, infer_data_name)
```
[the remark this answers is not in the document]
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:49]: Currently we only support ogbg-mol and CSV datasets. We can make it more flexible if we have more datasets in the future.

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    model_node_feat_size = cfg['model']['data_info']['node_feat_size']
+    model_edge_feat_size = cfg['model']['data_info']['edge_feat_size']
+    data_node_feat_size = data.node_feat_size
+    data_edge_feat_size = data.edge_feat_size
+    assert model_node_feat_size == data_node_feat_size, 'Expect the training data and inference \
+        data to have the same number of input node features, got {:d} and {:d}'.format(model_node_feat_size, data_node_feat_size)
+    assert model_edge_feat_size == data_edge_feat_size, 'Expect the training data and inference \
```
[the remark this answers is not in the document]
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:56]: Yes, then you are comparing None with an int.

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py (hunk 1 above)
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:23]: We only support binary classification for both training and inference. We might want to make it clear in the doc. However, there isn't much that can be done in the scripts.

Comment (jermainewang) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:23]: Besides doc update, I suggest also have a check of the label shape in the training script to make sure it is binary classification.

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:49]: Why only ogbg-mol? Is it because currently we only support binary classification?

File: dglgo/dglgo/apply_pipeline/graphpred/gen.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply dglgo/dglgo/apply_pipeline/graphpred/gen.py:23]: I don't get it. Why will it handle evaluation metrics? Shouldn't `dgl apply` be used when the prediction target/label is not known?

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py (hunk 1 above)
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:23]: `label shape` cannot tell you about that. For multi-label binary classification, the shape is `(N, T)`, where `T` is the number of tasks. For multi-class classification, the shape is `(N, C)`, where `C` is the number of classes.

File: dglgo/dglgo/apply_pipeline/graphpred/gen.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/gen.py:23]: This was discussed in our initial proposal that some labels might be available and users may want to evaluate some metrics. This is currently not implemented yet though. This naming is also consistent with `results` used by `dgl train`.

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:49]: Generally, this might require supporting alternative encoders/model architectures (e.g., ogbg-code2), loss functions & evaluation metrics (multi-class classification, regression), outer loop (cross validation), etc.

[the remark this answers is not in the document]
Comment (jermainewang) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:49]: I see. Then can users choose datasets other than ogbg-mol and CSV via commandline? If not, I'm ok with the current design.

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py (hunk 1 above)
Comment (jermainewang) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:23]: Got it. Then let's make it clear that this pipeline is for binary classification (in doc and in commandline help message as well).

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:49]: No, the valid options only include `ogbg-molhiv`, `ogbg-molpcba`, and `csv`.

File: dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py (hunk 1 above)
Comment (mufeili) [reply dglgo/dglgo/apply_pipeline/graphpred/graphpred.jinja-py:23]: updated

**Review by jermainewang — APPROVED** [review]
